cdap-api/src/main/java/co/cask/cdap/api/common/RuntimeArguments.java

```diff
@@ -74,33 +77,74 @@ private RuntimeArguments() {
+   * Arguments specified at the global level have lowest precedence.
+   *
    * @param scope The type of the scope
    * @param name The name of the scope, e.g. "myTable"
    * @param arguments the runtime arguments of the enclosing scope
    * @return a map that contains only the keys that start with &lt;scope>.&lt;name>., with that prefix removed.
    */
   public static Map<String, String> extractScope(Scope scope, String name, Map<String, String> arguments) {
-    String prefix = scope + "." + name + ".";
-    return extractPrefix(prefix, arguments);
-  }
-
-  /**
-   * Extract all arguments that start with a prefix, and removes that prefix.
-   * @param prefix The prefix to filter and remove
-   * @param arguments the runtime arguments to extract from
-   * @return a map that contains only the keys that start with the prefix, with the prefix removed.
-   */
-  public static Map<String, String> extractPrefix(String prefix, Map<String, String> arguments) {
     if (arguments == null || arguments.isEmpty()) {
       return arguments;
     }
+
+    String prefix = scope + DOT + name + DOT;
+    String wildCardPrefix = scope + DOT + STAR + DOT;
     Map<String, String> result = Maps.newHashMap();
+    Map<String, String> prefixMatchedArgs = Maps.newHashMap();
+    Map<String, String> wildCardPrefixMatchedArgs = Maps.newHashMap();
+    Map<String, String> noMatchArgs = Maps.newHashMap();
+
+    // Group the arguments into different categories based on wild card prefix match, named prefix match or no match
     for (Map.Entry<String, String> entry : arguments.entrySet()) {
-      if (entry.getKey().startsWith(prefix)) {
-        result.put(entry.getKey().substring(prefix.length()), entry.getValue());
+      if (entry.getKey().startsWith(wildCardPrefix)) {
+        // Argument is prefixed with "<scope>.*."
+        wildCardPrefixMatchedArgs.put(entry.getKey().substring(wildCardPrefix.length()), entry.getValue());
+      } else if (entry.getKey().startsWith(prefix)) {
+        // Argument is prefixed with "<scope>.<name>."
+        prefixMatchedArgs.put(entry.getKey().substring(prefix.length()), entry.getValue());
+      } else {
+        // Arguments specified for other scopes or without any scope
+        noMatchArgs.put(entry.getKey(), entry.getValue());
       }
     }
+
+    // Remove the arguments specified by other scopes. e.g. If the current scope is "MAPREDUCE.myMapReduce", remove
+    // the arguments specified with other scopes(except DATASET) such as "MAPREDUCE.yourMapReduce", "SPARK.somespark.".
+    // Note that DATASET scope requires special handling. If the current scope is "MAPREDUCE.myMapReduce", we still want
+    // to pass all the arguments scoped by DATASET. Arguments scoped by DATASET will only get filtered if current scope
+    // is DATASET.
+    Iterator<Map.Entry<String, String>> it = noMatchArgs.entrySet().iterator();
```

Comment:
You can actually have the logic here being down in the for loop above when determining whether to add to `noMatchArgs` or not. You can make the logic less clutter by moving it as a private static method. Something like:

`void addGlobalArgument(String key, String value, Scope scope, Map<String, String> result);`